File: backend/app/schemas/board.py
"""
게시판 관련 Pydantic 스키마
게시글, 댓글, 좋아요 등의 스키마 정의
"""
from __future__ import annotations
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


# ========== 게시글 신고 스키마 ==========

class PostReportCreate(BaseModel):
    """게시글 신고 요청 스키마"""
    reason: str = Field(..., description="신고 사유")
    details: Optional[str] = Field(None, description="상세 사유")

    class Config:
        json_schema_extra = {
            "example": {
                "reason": "욕설/비방",
                "details": "부적절한 언어를 사용했습니다."
            }
        }


# ========== 게시글 스키마 ==========

class PostCreate(BaseModel):
    """게시글 작성 요청 스키마"""
    title: str = Field(..., min_length=1, max_length=300, description="게시글 제목")
    content: str = Field(..., min_length=1, description="게시글 내용")
    category: str = Field(default="general", description="카테고리 (tip, question, free, general)")
    manual_tags: Optional[List[str]] = Field(default=None, description="사용자가 직접 추가한 태그")

    class Config:
        json_schema_extra = {
            "example": {
                "title": "피아노 연습 팁",
                "content": "오늘은 스케일 연습을 하면서...",
                "category": "tip",
                "manual_tags": ["연습", "피아노"]
            }
        }


class PostUpdate(BaseModel):
    """게시글 수정 요청 스키마"""
    title: Optional[str] = Field(None, min_length=1, max_length=300, description="게시글 제목")
    content: Optional[str] = Field(None, min_length=1, description="게시글 내용")
    category: Optional[str] = Field(None, description="카테고리 (tip, question, free, general)")
    manual_tags: Optional[List[str]] = Field(default=None, description="사용자가 직접 추가한 태그")

    class Config:
        json_schema_extra = {
            "example": {
                "title": "수정된 제목",
                "content": "수정된 내용",
                "category": "question",
                "manual_tags": ["질문"]
            }
        }


class PostAuthorResponse(BaseModel):
    """게시글 작성자 정보 응답 스키마"""
    user_id: int
    nickname: str
    profile_image_url: Optional[str] = None
    selected_achievement: Optional[AchievementResponse] = None  # 선택한 칭호 정보

    class Config:
        from_attributes = True


class PostResponse(BaseModel):
    """게시글 응답 스키마"""
    post_id: int
    user_id: int
    author: PostAuthorResponse
    title: str
    content: str
    category: str
    tags: Optional[List[str]] = None  # manual_tags (이전에 manual_tags로 저장된 태그)
    view_count: int
    like_count: int
    comment_count: int = 0  # 댓글 수 (대댓글 포함)
    is_liked: bool = False  # 현재 사용자가 좋아요를 눌렀는지 여부
    is_bookmarked: bool = False  # 현재 사용자가 북마크를 했는지 여부
    is_reported: bool = False  # 현재 사용자가 신고를 했는지 여부
    created_at: datetime
    updated_at: Optional[datetime] = None  # 수정된 적이 없으면 None

    class Config:
        from_attributes = True


class PostListResponse(BaseModel):
    """게시글 목록 응답 스키마"""
    posts: List[PostResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

    class Config:
        json_schema_extra = {
            "example": {
                "posts": [],
                "total": 0,
                "page": 1,
                "page_size": 20,
                "total_pages": 0
            }
        }


# ========== 댓글 스키마 ==========

class CommentCreate(BaseModel):
    """댓글 작성 요청 스키마"""
    content: str = Field(..., min_length=1, description="댓글 내용")
    parent_comment_id: Optional[int] = Field(None, description="부모 댓글 ID (답글인 경우)")

    class Config:
        json_schema_extra = {
            "example": {
                "content": "좋은 글 감사합니다!",
                "parent_comment_id": None
            }
        }


class CommentUpdate(BaseModel):
    """댓글 수정 요청 스키마"""
    content: str = Field(..., min_length=1, description="댓글 내용")

    class Config:
        json_schema_extra = {
            "example": {
                "content": "수정된 댓글 내용"
            }
        }


class CommentResponse(BaseModel):
    """댓글 응답 스키마"""
    comment_id: int
    post_id: int
    user_id: int
    author: PostAuthorResponse
    parent_comment_id: Optional[int] = None
    content: str
    like_count: int
    is_liked: bool = False  # 현재 사용자가 좋아요를 눌렀는지 여부
    replies: List['CommentResponse'] = []  # 답글 목록
    deleted_at: Optional[datetime] = None  # 삭제된 댓글인 경우 삭제 시간
    created_at: datetime
    updated_at: Optional[datetime] = None  # 수정된 적이 없으면 None

    class Config:
        from_attributes = True


# Forward reference 해결을 위한 import 및 model_rebuild
from app.schemas.achievements import AchievementResponse
PostAuthorResponse.model_rebuild()
CommentResponse.model_rebuild()


class CommentListResponse(BaseModel):
    """댓글 목록 응답 스키마"""
    comments: List[CommentResponse]
    total: int

    class Config:
        json_schema_extra = {
            "example": {
                "comments": [],
                "total": 0
            }
        }


# ========== 좋아요 스키마 ==========

class LikeResponse(BaseModel):
    """좋아요 응답 스키마"""
    is_liked: bool = Field(..., description="좋아요 상태 (true: 좋아요, false: 좋아요 취소)")
    like_count: int = Field(..., description="좋아요 개수")

    class Config:
        json_schema_extra = {
            "example": {
                "is_liked": True,
                "like_count": 10
            }
        }


class BookmarkResponse(BaseModel):
    """북마크 응답 스키마"""
    is_bookmarked: bool = Field(..., description="북마크 상태 (true: 추가됨, false: 삭제됨)")

    class Config:
        json_schema_extra = {
            "example": {
                "is_bookmarked": True
            }
        }


# ========== 메시지 응답 스키마 ==========

class MessageResponse(BaseModel):
    """일반 메시지 응답 스키마"""
    message: str

    class Config:
        json_schema_extra = {
            "example": {
                "message": "작업이 완료되었습니다."
            }
        }

File: frontend/src/services/api/board.ts
import apiClient from './client';

// 게시판 API 타입 정의
export interface PostAuthor {
  user_id: number;
  nickname: string;
  profile_image_url?: string | null;
  selected_achievement?: {
    achievement_id: number;
    title: string;
    description?: string | null;
    icon_url?: string | null;
  } | null;
}

export interface Post {
  post_id: number;
  user_id: number;
  author: PostAuthor;
  title: string;
  content: string;
  category: string;
  tags?: string[] | null;  // manual_tags (이전에 manual_tags로 저장된 태그)
  view_count: number;
  like_count: number;
  comment_count: number;
  is_liked: boolean;
  is_bookmarked: boolean;
  created_at: string;
  updated_at?: string | null;  // 수정된 적이 없으면 null
}

export interface PostListResponse {
  posts: Post[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface PostCreate {
  title: string;
  content: string;
  category?: string;
  manual_tags?: string[];
}

export interface PostUpdate {
  title?: string;
  content?: string;
  category?: string;
  manual_tags?: string[];
}

export interface PostReportCreate {
  reason: string;
  details?: string;
}

export interface CommentAuthor {
  user_id: number;
  nickname: string;
  profile_image_url?: string | null;
  selected_achievement?: {
    achievement_id: number;
    title: string;
    description?: string | null;
    icon_url?: string | null;
  } | null;
}

export interface PostComment {
  comment_id: number;
  post_id: number;
  user_id: number;
  author: CommentAuthor;
  parent_comment_id?: number | null;
  content: string;
  like_count: number;
  is_liked: boolean;
  replies: PostComment[];
  deleted_at?: string | null;  // 삭제된 댓글인 경우 삭제 시간
  created_at: string;
  updated_at?: string | null;  // 수정된 적이 없으면 null
}

export interface CommentListResponse {
  comments: PostComment[];
  total: number;
}

export interface CommentCreate {
  content: string;
  parent_comment_id?: number | null;
}

export interface CommentUpdate {
  content: string;
}

export interface LikeResponse {
  is_liked: boolean;
  like_count: number;
}

export interface MessageResponse {
  message: string;
}

export const boardApi = {
  /**
   * 게시글 목록 조회
   */
  getPosts: async (params?: {
    page?: number;
    page_size?: number;
    category?: string;
    tag?: string;
    search?: string;
    author_id?: number;
    bookmarked_only?: boolean;
  }): Promise<PostListResponse> => {
    const response = await apiClient.get<PostListResponse>('/board/posts', { params });
    return response.data;
  },

  /**
   * 게시글 작성
   */
  createPost: async (data: PostCreate): Promise<Post> => {
    const response = await apiClient.post<Post>('/board/posts', data);
    return response.data;
  },

  /**
   * 게시글 상세 조회
   */
  getPost: async (postId: number): Promise<Post> => {
    const response = await apiClient.get<Post>(`/board/posts/${postId}`);
    return response.data;
  },

  /**
   * 게시글 수정
   */
  updatePost: async (postId: number, data: PostUpdate): Promise<Post> => {
    const response = await apiClient.put<Post>(`/board/posts/${postId}`, data);
    return response.data;
  },

  /**
   * 게시글 삭제
   */
  deletePost: async (postId: number): Promise<MessageResponse> => {
    const response = await apiClient.delete<MessageResponse>(`/board/posts/${postId}`);
    return response.data;
  },

  /**
   * 게시글 좋아요 토글
   */
  togglePostLike: async (postId: number): Promise<LikeResponse> => {
    const response = await apiClient.post<LikeResponse>(`/board/posts/${postId}/likes`);
    return response.data;
  },

  /**
   * 게시글 북마크 토글
   */
  togglePostBookmark: async (postId: number): Promise<{ is_bookmarked: boolean }> => {
    const response = await apiClient.post<{ is_bookmarked: boolean }>(`/board/posts/${postId}/bookmarks`);
    return response.data;
  },

  /**
   * 게시글 신고
   */
  reportPost: async (postId: number, data: PostReportCreate): Promise<MessageResponse> => {
    const response = await apiClient.post<MessageResponse>(`/board/posts/${postId}/report`, data);
    return response.data;
  },

  /**
   * 댓글 목록 조회
   */
  getComments: async (postId: number): Promise<CommentListResponse> => {
    const response = await apiClient.get<CommentListResponse>(`/board/posts/${postId}/comments`);
    return response.data;
  },

  /**
   * 댓글 작성
   */
  createComment: async (postId: number, data: CommentCreate): Promise<PostComment> => {
    const response = await apiClient.post<PostComment>(`/board/posts/${postId}/comments`, data);
    return response.data;
  },

  /**
   * 댓글 수정
   */
  updateComment: async (commentId: number, data: CommentUpdate): Promise<PostComment> => {
    const response = await apiClient.put<PostComment>(`/board/comments/${commentId}`, data);
    return response.data;
  },

  /**
   * 댓글 삭제
   */
  deleteComment: async (commentId: number): Promise<MessageResponse> => {
    const response = await apiClient.delete<MessageResponse>(`/board/comments/${commentId}`);
    return response.data;
  },

  /**
   * 댓글 좋아요 토글
   */
  toggleCommentLike: async (commentId: number): Promise<LikeResponse> => {
    const response = await apiClient.post<LikeResponse>(`/board/comments/${commentId}/likes`);
    return response.data;
  },
};

